cli/internal/daemon/daemon.go

```diff
@@ -0,0 +1,324 @@
+	root := getDaemonFileRoot(repoRoot)
+	return root.Join("turbod.sock")
+}
+
+func getPidFile(repoRoot fs.AbsolutePath) fs.AbsolutePath {
+	root := getDaemonFileRoot(repoRoot)
+	return root.Join("turbod.pid")
+}
+
+// logError logs an error and outputs it to the UI.
+func (d *daemon) logError(err error) {
+	d.logger.Error("error", err)
+}
+
+// we're only appending, and we're creating the file if it doesn't exist.
+// we do not need to read the log file.
+var _logFileFlags = os.O_WRONLY | os.O_APPEND | os.O_CREATE
+
+func getCmd(config *config.Config, output cli.Ui, signalWatcher *signals.Watcher) *cobra.Command {
+	var idleTimeout time.Duration
+	cmd := &cobra.Command{
+		Use:           "turbo daemon",
+		Short:         "Runs turbod",
+		SilenceUsage:  true,
+		SilenceErrors: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logFilePath, err := getLogFilePath(config.Cwd)
+			if err != nil {
+				return err
+			}
+			if err := logFilePath.EnsureDir(); err != nil {
+				return err
+			}
+			logFile, err := logFilePath.OpenFile(_logFileFlags, 0644)
+			if err != nil {
+				return err
+			}
+			defer func() { _ = logFile.Close() }()
+			logger := hclog.New(&hclog.LoggerOptions{
+				Output: io.MultiWriter(logFile, os.Stdout),
+				Level:  hclog.Debug,
+				Color:  hclog.ColorOff,
+				Name:   "turbod",
+			})
+			ctx := cmd.Context()
+			d := &daemon{
+				logger:     logger,
+				repoRoot:   config.Cwd,
+				timeout:    idleTimeout,
+				reqCh:      make(chan struct{}),
+				timedOutCh: make(chan struct{}),
+			}
+			turboServer, err := server.New(d.logger.Named("rpc server"), config.Cwd, config.TurboVersion, logFilePath)
+			if err != nil {
+				d.logError(err)
+				return err
+			}
+			defer func() { _ = turboServer.Close() }()
+			err = d.runTurboServer(ctx, turboServer, signalWatcher)
+			if err != nil {
+				d.logError(err)
+				return err
+			}
+			return nil
+		},
+	}
+	cmd.Flags().DurationVar(&idleTimeout, "idle-time", 2*time.Hour, "Set the idle timeout for turbod")
+	addDaemonSubcommands(cmd, config, output)
+	return cmd
+}
+
+func addDaemonSubcommands(cmd *cobra.Command, config *config.Config, output cli.Ui) {
+	addStatusCmd(cmd, config, output)
+	addStartCmd(cmd, config, output)
+	addStopCmd(cmd, config, output)
+	addRestartCmd(cmd, config, output)
+}
+
+var errInactivityTimeout = errors.New("turbod shut down from inactivity")
+
+// debounceServers attempts to ensure that only one daemon is running from the given pid file path
+// at a time.
+func (d *daemon) debounceServers(pidPath fs.AbsolutePath) (lockfile.Lockfile, error) {
+	lockFile, err := lockfile.New(pidPath.ToString())
+	if err != nil {
+		return "", err
+	}
+	if err := lockFile.TryLock(); err != nil {
+		return "", err
+	}
+	return lockFile, nil
+
+}
+
+type rpcServer interface {
+	Register(grpcServer server.GRPCServer)
+}
+
+func (d *daemon) runTurboServer(parentContext context.Context, rpcServer rpcServer, signalWatcher *signals.Watcher) error {
```

Comment:
I want to walk through this function in great detail on our next call to make sure I understand it perfectly.